mods/win7-alttab-loader.wh.cpp

```diff
@@ -16,568 +16,332 @@
-            return 0;
-        
-        GetDesktopText(lpBuffer);
-
-        int len = wcslen(lpBuffer);
-        for (int i = 0; i < len; i++)
-            if (lpBuffer[i] == L'&')
-                lpBuffer[i] = L'\u200E';
-        
-        return len;
-    }
+        PROPERTYKEY pKey;
+        pKey.fmtid = GUID_AppUserModelIdProperty;
+        pKey.pid = 5;
+        PROPVARIANT prop;
+        ZeroMemory(&prop, sizeof(PROPVARIANT));
+        pps->GetValue(pKey, &prop);
+        pps->Release();
+        if (prop.bstrVal)
+        {
+            SHCreateItemInKnownFolder(
+                FOLDERID_AppsFolder,
+                KF_FLAG_DONT_VERIFY,
+                prop.bstrVal,
+                IID_PPV_ARGS(&psiif)
+            );
+            if (psiif)
+            {
+                int szIcon = GetSystemMetrics(SM_CXICON);
+
+                SIZE size;
+                size.cx = szIcon;
+                size.cy = szIcon;
+                HBITMAP hBitmap;
+                hr = psiif->GetImage(
+                    size,
+                    flags,
+                    &hBitmap
+                );
+                if (SUCCEEDED(hr))
+                {
+                    // Easiest way to get an HICON from an HBITMAP
+                    // I have turned the Internet upside down and was unable to find this
+                    // Only a convoluted example using GDI+
+                    // This is from the disassembly of StartIsBack/StartAllBack
+                    HIMAGELIST hImageList = ImageList_Create(size.cx, size.cy, ILC_COLOR32, 1, 0);
+                    if (ImageList_Add(hImageList, hBitmap, NULL) != -1)
+                    {
+                        result = ImageList_GetIcon(hImageList, 0, 0);
+                        ImageList_Destroy(hImageList);
 
-    return LoadStringW(hInstance, uId, lpBuffer, cchBufferMax);
+                        DeleteObject(hBitmap);
+                        psiif->Release();
+                    }
+                    DeleteObject(hBitmap);
+                }
+                psiif->Release();
+            }
+        }
+    }
+    return result;
 }
 
-BOOL PostMessageW_Hook(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
+/* Support UWP icons */
+SENDASYNCPROC CWindowInfo__GetIconProc_orig;
+void WINAPI CWindowInfo__GetIconProc_hook(
+    HWND     hWnd,
+    UINT     uMsg,
+    UINT_PTR upParam,
+    LRESULT  lResult
+)
 {
-    if (hWnd == FindWindowW(L"Shell_TrayWnd", NULL) && uMsg == 0x5B7 && wParam == 0 && lParam == 0)
+    if (NULL == lResult && IsShellFrameWindow(hWnd))
     {
-        return PostMessageW(hWnd, WM_COMMAND, 407, 0);
+        lResult = (LRESULT)GetUWPIcon(hWnd);
     }
-
-    return PostMessageW(hWnd, uMsg, wParam, lParam);
+    CWindowInfo__GetIconProc_orig(hWnd, uMsg, upParam, lResult);
 }
 
-HTHEME OpenThemeData_Hook(HWND hWnd, LPCWSTR pszClassList)
-{
-    if (wcscmp(pszClassList, L"AltTab") == 0)
+const WindhawkUtils::SYMBOL_HOOK altTabDllHooks[] = {

```

Comment:
You're challenging my symbol cache system :)

> "altTab.Dll" is not recognized as a Windows file name

That's because it's not available in Win10 and newer. Interesting that Microsoft still provides symbols for Win7 binaries. I wonder how long they keep it and what about e.g. WinXP.

From https://en.wikipedia.org/wiki/Windows_7:
> Mainstream support ended
> [...]
> Exceptions exist until at most October 8, 2024, see [§ Support lifecycle](https://en.wikipedia.org/wiki/Windows_7#Support_lifecycle) for details.

They might to decide removing these symbols at some point.